torch/csrc/jit/python/pybind_utils.h

```diff
@@ -476,6 +476,17 @@ inline void guardAgainstNamedTensor(const T& var) {
       "workaround please drop names via `tensor = tensor.rename(None)`.");
 }
 
+// NB: When passing py::object as the first argument, the call site needs to
+// keep the py::objec alive until this function returns.
+//
+// Note [jit::toIValue barrow py::object refcnt]
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Comment:
Hey @ezyang, I am not sure if my understanding is correct here. Is it true that if we pass a `py::object` to `toIValue`, we need to make sure that this `py::object` is alive until `toIValue` returns, because py::handle here does not inc the refcnt?

Reply:
Yes this is correct, because `toIValue` mostly used as function argument conversion, and it is almost certain that the caller manages the lifetime of the `py::handle` passed in.

Reply:
Might be a naive question, but what's the problem with changing `toIValue` to call `inc_ref` to keep the underlying object alive for the duration of the function call, or even just use `py::object` here? This would make it a lot easier for callers and potentially avoid bugs/flakiness. Are we concerned about the performance implications?

Reply:
Don't we care that the value of the `py::object` is persisted for the duration of `markComplete()`, not just `jit::toIValue()`?

Reply:
If below is the code that runs when passing a py::object to py::handle, it actually inc the refcnt?

https://github.com/pybind/pybind11/blob/a86ac538f59db7af596ee192f42b9f39ec8039ed/include/pybind11/cast.h#L1623-L1625

```c++
struct pyobject_caster {
    template <typename T = type, enable_if_t<std::is_same<T, handle>::value, int> = 0>
    bool load(handle src, bool /* convert */) { value = src; return static_cast<bool>(value); }

    template <typename T = type, enable_if_t<std::is_base_of<object, T>::value, int> = 0>
    bool load(handle src, bool /* convert */) {
        if (!isinstance<type>(src))
            return false;
        value = reinterpret_borrow<type>(src);
        return true;
    }

    static handle cast(const handle &src, return_value_policy /* policy */, handle /* parent */) {
        return src.inc_ref();
    }
    PYBIND11_TYPE_CASTER(type, handle_type_name<type>::name);
};
```

Reply:
> Don't we care that the value of the py::object is persisted for the duration of markComplete(), not just jit::toIValue()?

`jit::toIValue` should have created a copy of the `py::object` in the cast below. So as long as the IValue is alive, the py::object would also be alive I think?

https://github.com/pytorch/pytorch/blob/ee1ddcef8d2994b59c9d7aeae8f0dd455ee6e469/torch/csrc/jit/python/python_ivalue.h#L17-L21

Reply:
> Might be a naive question, but what's the problem with changing toIValue to call inc_ref to keep the underlying object alive for the duration of the function call, or even just use py::object here? This would make it a lot easier for callers and potentially avoid bugs/flakiness. Are we concerned about the performance implications?

Not sure what are the reason for using py::handle here. @wanchaol is there any specific reason for this?

Reply:
I don't remember if there's a specific reason to keep a `py::handle`, merely to avoid redundant inc_refs as arguments are suppose to be alive when you pass to a function.

cc @zdevito who might have more context on this.

Reply:
Say we have
```cpp
py::object o;

py::handle aa(o);  // copy constructor
py::handle aaa = o;  // copy constructor, as it's also initialization.
```
The constructor uses here is the synthesized default copy-constructor, `py::handle(const py::handle&)`.
`py::object` is statically casted to `const py::handle&` type as it's a subtype.
Also the lifetime of the RValue(temporary) `py::object` returned by `toPyPObject(..)` is prolonged by binding to the const reference (`const py::handle&`). 

Similar for,
```cpp
py::handle bb; // The empty constructor of py::handle is called.
bb = o;  // copy assign operator
```
The synthesized default copy-assign-operator, `py::handle::operator=(const py::handle&)`.